src/Server/TCPHandler.cpp

```diff
@@ -845,6 +889,15 @@ void TCPHandler::receiveQuery()
     state.is_empty = false;
     readStringBinary(state.query_id, *in);
 
+    if (hello_credentials)
```

Comment:
If the interserver secret is enabled then why do you want to transfer the password? The server shouldn't use the password, it should use the interserver secret for authentication. So you're sending the password while the server doesn't need it.

And if the interserver secret is disabled you're making the things worse because now you're going to keep the password plaintext in memory for a while.

Reply:
I think it could be better to move the authentication check back into `TCPHandler::receiveHello()` to work as usual if the interserver secret is disabled. And if the interserver secret is enabled the client could send some reserved string like "INTERSERVER SECRET IS ENABLED" instead of a user name and an empty string instead of a password inside the standard hello packet. And if the server is new enough it will understand that such user name means that the server is going to receive an extra information with cluster name and salt. In this case you won't have to keep the credentials in memory.

Reply:
>I think it could be better to move the authentication check back into TCPHandler::receiveHello() to work as usual if the interserver secret is disabled. And if the interserver secret is enabled the client could send some reserved string like "INTERSERVER SECRET IS ENABLED" instead of a user name and an empty string instead of a password inside the standard hello packet.

User is used for hashing.

> And if the server is new enough it will understand that such user name means that the server is going to receive an extra information with cluster name and salt. In this case you won't have to keep the credentials in memory.

I went another way, for inter-server secret, I sent empty user (anyway it is not allowed) and the real user goes in the password, and if the user is not empty then it calls setUser just in receiveHello, otherwise it will call setUserWithoutCheckingPassword later while receiving the query.

This has some downsides, if the remote shard does not support this then the error will be about empty user.

P.S. I did not added any checks right now (i.e. to avoid `Cluster` package, if setUser is called), just to show the idea (this PR already contains lots of behavior changes and merge commits, that will be not that easy to rebase)

Reply:
I meant the reserved name like "INTERSERVER SECRET IS ENABLED" (not just empty name) is good because
1) if the server is old so it doesn't support the interserver secret then you will get a quite descriptive error message anyway like `INTERSERVER SECRET IS ENABLED : authentication error`
2) you can send cluster info data right after the standard hello package (i.e. in the same package), and the server will understand from the reserved user name it should read extra data. And sending extra data in the standard package won't break the compatibility in this case cause if the server is old it will stop with authentication error anyway.

Empty name can mean anything, maybe an error, but a reserved name can have a definitive meaning.

Reply:
I like this idea, done (although looks a little bit tricky, i.e. basically it does not depends from the revision except for query hash now)